### PR TITLE
Uses corrected two pass algorithm in theano.tensor.var

### PR DESCRIPTION
**Pull request by kumarkrishna**
Reference #4651 
Adds argument ```corrected``` to maintain backward compatibility.

### CODE REVIEW

**Comment by slefrancois** [comment]
test this please
we're testing a jenkins bot

File: theano/tensor/basic.py [hunk 1]
```diff
@@ -3167,6 +3170,12 @@ def var(input, axis=None, keepdims=False):
 
     # return the mean sqr
     v = mean((centered_input ** 2), axis, keepdims=keepdims)
+
+    # use 'corrected_two_pass' algorithm
```
Comment (ballasn) [inline theano/tensor/basic.py]: you should add a reference to the tech report http://www.cs.yale.edu/publications/techreports/tr222.pdf

File: theano/tensor/tests/test_basic.py [hunk 2]
```diff
@@ -6332,6 +6332,8 @@ def test_var():
     f = function([a], var(a, axis=2))
     assert numpy.allclose(numpy.var(a_val, axis=2), f(a_val))
 
+    f = function([a], var(a, corrected=True))
+    assert numpy.allclose(numpy.var(a_val), f(a_val))
```
Comment (ballasn) [inline theano/tensor/tests/test_basic.py]: I don't think numpy.var uses the corrected two pass algorithm so you probably to implement the corrected two path algorithm as well for the test.

**Comment by nouiz** [comment]
@kumarkrishna Can you finish this PR? I think not much is left and it would be great to close this.

thanks.

**Comment by kumarkrishna** [comment]
@nouiz Updated the PR with more tests, fixed merge conflicts. I have also updated the implementation to take into consideration the ```ddof``` parameter. I think the patch is ready for review and merge.

File: theano/tensor/basic.py [hunk 3]
```diff
@@ -3179,11 +3179,16 @@ def var(input, axis=None, ddof=0, keepdims=False):
...
         the result will broadcast correctly against the original tensor.
+    corrected : bool
+        If this is set to True, the 'corrected_two_pass' algorithm is
+        used to compute the variance.
+        Refer : http://www.cs.yale.edu/publications/techreports/tr222.pdf
 
     Notes
     -----
-    It uses the two-pass algorithm for more stable results.
+    By default, uses the two-pass algorithm for more stable results.
```
Comment (nouiz) [inline theano/tensor/basic.py]: It don't use it by default. Tell that it optionally support that.

File: theano/tensor/var.py [hunk 4]
```diff
@@ -643,9 +643,10 @@ def mean(self, axis=None, dtype=None, keepdims=False, acc_dtype=None):
                                         dtype=dtype, keepdims=keepdims,
                                         acc_dtype=acc_dtype)
 
-    def var(self, axis=None, keepdims=False):
+    def var(self, axis=None, keepdims=False, corrected=False):
```
Comment (nouiz) [inline theano/tensor/var.py]: add the ddof parameter at the same time here.

File: theano/tensor/basic.py (hunk 3 above)
Comment (kumarkrishna) [inline theano/tensor/basic.py]: > By default, uses the two-pass algorithm

This refers to the usual two-pass algorithm used default for calculating the variance, not the corrected version. I will rewrite to make it more clear.

**Comment by ballasn** [comment]
pr looks good to me, @nouiz  I think we can merge thanks,

**Comment by ballasn** [comment]
Merged, thanks!

**Pull request merged**